crypto_stream_chacha20.js

```diff
@@ -0,0 +1,214 @@
+    'k should be crypto_stream_chacha20_ietf_KEYBYTES')
+
+  return new Chacha20(n, k)
+}
+
+function Chacha20 (n, k, counter) {
+  assert(k.byteLength === exports.crypto_stream_chacha20_ietf_KEYBYTES)
+  assert(n.byteLength === exports.crypto_stream_chacha20_NONCEBYTES ||
+    n.byteLength === exports.crypto_stream_chacha20_ietf_NONCEBYTES)
+
+  if (!counter) counter = 0
+  assert(counter < Number.MAX_SAFE_INTEGER)
+
+  this.finalized = false
+  this.pos = 0
+  this.state = new Uint32Array(16)
+
+  for (let i = 0; i < 4; i++) this.state[i] = constant[i]
+  for (let i = 0; i < 8; i++) this.state[4 + i] = k.readUInt32LE(4 * i)
+
+  this.state[12] = counter & 0xffffffff
+
+  if (n.byteLength === 8) {
+    this.state[13] = counter >> 32
+    this.state[14] = n.readUInt32LE(0)
+    this.state[15] = n.readUInt32LE(4)
+  } else {
+    this.state[13] = n.readUInt32LE(0)
+    this.state[14] = n.readUInt32LE(4)
+    this.state[15] = n.readUInt32LE(8)
+  }
+
+  return this
+}
+
+Chacha20.prototype.update = function (output, input) {
+  assert(!this.finalized, 'cipher finalized.')
+  assert(output.byteLength >= input.byteLength,
+    'output cannot be shorter than input.')
+
+  var len = input.length
+  var offset = this.pos % 64
+  this.pos += len
+
+  // input position
+  var j = 0
+
+  var keyStream = chacha20_block(this.state)
+
+  // try to finsih the current block
+  while (offset > 0 && len > 0) {
+    output[j] = input[j++] ^ keyStream[offset]
+    offset = (offset + 1) & 0x3f
+    if (!offset) this.state[12]++
+    len--
+  }
+
+  // encrypt rest block at a time
+  while (len > 0) {
+    keyStream = chacha20_block(this.state)
+
+    // less than a full block remaining
+    if (len < 64) {
+      for (let i = 0; i < len; i++) {
+        output[j] = input[j++] ^ keyStream[offset++]
+        offset &= 0x3f
+      }
+
+      return
+    }
+
+    for (; offset < 64;) {
+      output[j] = input[j++] ^ keyStream[offset++]
+    }
+
+    this.state[12]++
+    offset = 0
+    len -= 64
+  }
+}
+
+Chacha20.prototype.final = function () {
+  this.finalized = true
+}
+
+module.exports.keystream = function (output, key, nonce, counter) {
+  var c = new Chacha20(key, nonce, counter)
+  c.update(output, Buffer.alloc(output.length))
+  c.final()
+}
+
+function chacha20_block (state) {
+  var workingState = new Uint32Array(16)
+  for (let i = 16; i--;) workingState[i] = state[i]
+
+  for (let i = 0; i < 20; i += 2) {
+    QR(workingState, 0, 4, 8, 12) // column 0
+    QR(workingState, 1, 5, 9, 13) // column 1
+    QR(workingState, 2, 6, 10, 14) // column 2
+    QR(workingState, 3, 7, 11, 15) // column 3
+
+    QR(workingState, 0, 5, 10, 15) // diagonal 1 (main diagonal)
+    QR(workingState, 1, 6, 11, 12) // diagonal 2
+    QR(workingState, 2, 7, 8, 13) // diagonal 3
+    QR(workingState, 3, 4, 9, 14) // diagonal 4
+  }
+
+  for (let i = 0; i < 16; i++) {
+    workingState[i] += state[i]
+  }
+
+  return Buffer.from(workingState.buffer)
```

Comment:
from(workingState.buffer, workingState.byteOffset, workingState.byteLength)

Reply:
patched in fd7372d